### DISCUSSION TITLE
Dynamic items in Annotated list.

### DISCUSSION THREAD

**Discussion by m4dm4rtig4n** [fastapi/fastapi#11442] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum
from typing import Annotated, Optional

from fastapi import APIRouter, HTTPException, Path
from pydantic import BaseModel

from models.database import DATABASE
from models.supplier import Supplier

ROUTER = APIRouter(
    tags=["Fournisseur"],
)

class TempSupplierList(str, Enum):
    """TempSupplierList enum."""

    pass


supplier_dict = {}
for supplier in DATABASE.list_supplier():
    supplier_dict[supplier.name] = supplier.name.upper()
SupplierList = TempSupplierList("SupplierList", supplier_dict)

@ROUTER.get(
    "/supplier/{supplier_name}",
    summary="Remonte les informations d'un fournisseur d'énergie.",
)
@ROUTER.get(
    "/supplier/{supplier_name}/",
    include_in_schema=False,
)
async def get_supplier(
    supplier_name: Annotated[
        SupplierList,
        Path(description="Liste des fournisseurs"),
    ],
    enable: Optional[bool] = None,
):
    """Get supplier."""
    supplier = Supplier(supplier_name).supplier_get(enable)
    if hasattr(supplier, "error") and not supplier.error:
        return supplier.item
    print(supplier.status_code, supplier.detail)
    raise HTTPException(status_code=int(supplier.status_code), detail=supplier.detail)
```

### Description

Hello,

I try to create dynamic SupplierList with data retreive dynamically from Database.
I have other endpoint "create_supplier", where end user can create other supplier and with this declaration SupplierList  is not update without app reload.

It's possible to have "SupplierList" reload dynamically ?

Thx

### Operating System

Linux

### FastAPI Version

0.88.0

### Pydantic Version

1.10.15

### Python Version

3.12.1

**Comment by JavierSanchezCastro** [comment]
Since you're creating SupplierList once, if you want to see the new suppliers after launching your application, you need to call the database again to get the new data. Therefore, if you don't want to call every time, you can consider implementing some caching strategy in your service.

**Comment by YuriiMotov** [comment, marked as answer]
Do you need these options to be listed in the openapi schema (Swagger docs)? By default this schema will be created once after you run app and request this schema, and you will always see constant list of options until you reload the app. Although, you can modify code to refresh openapi schema every time you update list of options.

Since you can run your app on multiple workers, consider using distributed cache (e.g. Redis) to store list of items or notifying other workers about changes.

Also, you should understand that for type checkers the use of dynamic enums will be treated as an error.


I would just check this condition in the endpoint and raise HTTPException if the value is not in the list of allowed options. In this case you will not see the list of allowed options in the docs (OpenAPI schema), but it's Ok for dynamic lists.
And if you need to do this check in different endpoints, then just create the dependency and move this logic there. In the endpoints use this dependency

**Reply by VaishnaviPandey02** [reply]
how can we use the dependency in the endpoints to get the dynamic dropdown on my swagger UI ?please explain with some example
